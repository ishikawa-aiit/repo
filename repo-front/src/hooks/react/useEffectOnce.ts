import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
