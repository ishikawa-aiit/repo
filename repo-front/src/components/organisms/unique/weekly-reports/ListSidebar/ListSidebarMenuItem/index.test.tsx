import { render, screen } from '@testing-library/react'
import React from 'react'

import { ListSidebarMenuItem } from '.'

const testMenuItem = {
  link: '0',
  name: 'testName0',
  isCurrent: false,
}

jest.mock('next/link', () => {
  return ({ children, ...rest }: { children: React.ReactElement }) => {
    return React.cloneElement(children, { ...rest })
  }
})

describe('ListSidebarMenuItem', () => {
  it('正常にレンダリングされている', () => {
    const { name } = testMenuItem
    const { container } = render(<ListSidebarMenuItem {...testMenuItem} />)

    expect(screen.getByText(name)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('指定のパスとリンクのhref属性が一致する', () => {
    const { link } = testMenuItem
    const { container } = render(<ListSidebarMenuItem {...testMenuItem} />)

    expect(screen.getByRole('link')).toHaveAttribute('href', `#${link}`)
    expect(container).toMatchSnapshot()
  })
})

describe('props:isCurrent', () => {
  it('trueの場合、リンクがカレントとなる', () => {
    const { container } = render(<ListSidebarMenuItem {...testMenuItem} isCurrent={true} />)

    expect(screen.getByTestId('ListSidebarCurrentMenuItem')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('falseの場合、リンクがカレントではない', () => {
    const { container } = render(<ListSidebarMenuItem {...testMenuItem} />)

    expect(screen.getByTestId('ListSidebarMenuItem')).toBeInTheDocument()
    expect(screen.queryByTestId('ListSidebarCurrentMenuItem')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
