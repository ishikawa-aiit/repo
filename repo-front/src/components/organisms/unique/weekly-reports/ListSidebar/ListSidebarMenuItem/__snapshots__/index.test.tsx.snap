// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListSidebarMenuItem 指定のパスとリンクのhref属性が一致する 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: primary.100;
}

<div>
  <a
    class="emotion-0"
    data-cy="ListSidebarMenuItem"
    data-testid="ListSidebarMenuItem"
    href="#0"
  >
    testName0
  </a>
</div>
`;

exports[`ListSidebarMenuItem 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: primary.100;
}

<div>
  <a
    class="emotion-0"
    data-cy="ListSidebarMenuItem"
    data-testid="ListSidebarMenuItem"
    href="#0"
  >
    testName0
  </a>
</div>
`;

exports[`props:isCurrent falseの場合、リンクがカレントではない 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: primary.100;
}

<div>
  <a
    class="emotion-0"
    data-cy="ListSidebarMenuItem"
    data-testid="ListSidebarMenuItem"
    href="#0"
  >
    testName0
  </a>
</div>
`;

exports[`props:isCurrent trueの場合、リンクがカレントとなる 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

<div>
  <a
    aria-current="page"
    class="emotion-0"
    data-cy="ListSidebarCurrentMenuItem"
    data-testid="ListSidebarCurrentMenuItem"
    href="#0"
  >
    testName0
  </a>
</div>
`;
